lib/adapters/http.js

```diff
@@ -15,158 +15,169 @@ var createError = require('../core/createError');
-    options.port = config.proxy.port;
-    options.path = parsed.protocol + '//' + parsed.hostname + options.path;
-  }
-
-  var transport;
-  if (config.maxRedirects === 0) {
-    transport = parsed.protocol === 'https:' ? https : http;
-  } else {
-    if (config.maxRedirects) {
-      options.maxRedirects = config.maxRedirects;
+    if (data && !utils.isStream(data)) {
+      if (utils.isArrayBuffer(data)) {
+        data = new Buffer(new Uint8Array(data));
+      } else if (utils.isString(data)) {
+        data = new Buffer(data, 'utf-8');
+      } else {
+        return reject(createError(
+          'Data after transformation must be a string, an ArrayBuffer, or a Stream',
+          config
+        ));
+      }
+
+      // Add Content-Length header if data exists
+      headers['Content-Length'] = data.length;
     }
-    transport = parsed.protocol === 'https:' ? httpsFollow : httpFollow;
-  }
-
-  // Create the request
-  var req = transport.request(options, function handleResponse(res) {
-    if (aborted) return;
-
-    // Response has been received so kill timer that handles request timeout
-    clearTimeout(timer);
-    timer = null;
-
-    // uncompress the response body transparently if required
-    var stream = res;
-    switch (res.headers['content-encoding']) {
-    /*eslint default-case:0*/
-    case 'gzip':
-    case 'compress':
-    case 'deflate':
-      // add the unzipper to the body stream processing pipeline
-      stream = stream.pipe(zlib.createUnzip());
-
-      // remove the content-encoding in order to not confuse downstream operations
-      delete res.headers['content-encoding'];
-      break;
+
+    // HTTP basic authentication
+    var auth = undefined;
+    if (config.auth) {
+      var username = config.auth.username || '';
+      var password = config.auth.password || '';
+      auth = username + ':' + password;
     }
 
-    var response = {
-      status: res.statusCode,
-      statusText: res.statusMessage,
-      headers: res.headers,
-      config: config,
-      request: req
+    // Parse url
+    var parsed = url.parse(config.url);
+    if (!auth && parsed.auth) {
+      var urlAuth = parsed.auth.split(':');
+      var urlUsername = urlAuth[0] || '';
+      var urlPassword = urlAuth[1] || '';
+      auth = urlUsername + ':' + urlPassword;
+    }
+    var options = {
+      hostname: parsed.hostname,
+      port: parsed.port,
+      path: buildURL(parsed.path, config.params, config.paramsSerializer).replace(/^\?/, ''),
+      method: config.method,
+      headers: headers,
+      agent: config.agent,
+      auth: auth
     };
 
-    if (config.responseType === 'stream') {
-      response.data = stream;
-      settle(resolve, reject, response);
+    var proxy = config.proxy;
+    if (!proxy) {
+      var proxyEnv = parsed.protocol.slice(0, -1) + '_proxy';
+      var proxyUrl = process.env[proxyEnv] || process.env[proxyEnv.toUpperCase()];
+      if (proxyUrl) {
+        proxy = url.parse(proxyUrl);
```

Comment:
I think we should use the `hostname` property of the URL object rather than `host`. The latter contains the host name *and* the port (e.g. 'foo.com:8080'), which is not what should be passed to `http.request` (see https://nodejs.org/api/http.html#http_http_request_options_callback).

```javascript
var parsedProxyUrl = url.parse(proxyUrl);
proxy = {
  host: parsedProxyUrl.hostname,
  port: parsedProxyUrl.port
};
```